### PR TITLE
Change the dbgshim launch handshake back

### PR DESCRIPTION
**Pull request by mikem8361**
The debugger side now creates the name semaphores like before and the following is used to fix the race conditions.

1) Add a global dword to coreclr and statically initialize it 0. Call it something like "g_coreclrProcessIsReady".
2) When the coreclr has finished the dac table initialization (the same place the continue semaphore was created before), set this global to the process's PID.
3) Export this field public from the coreclr module like the dac globals table (g_dacTable).
4) Add to the xplat build logic that puts the new g_corclrProcessIsReady global's RVA in the dactablerva.h include (along with DAC_TABLE_RVA define).
5) Add a simple public function on the dac module to return this new RVA.
6) In dbgshim, once the coreclr module is loaded and we know the path of the dac module, load it and call this function to get the RVA. If the function doesn't exist, this means older version of dac/coreclr modules so we may have races.
7) Use ReadProcessMemory of the coreclr module base + RVA in dbgshim to read g_coreclrProcessIsReady and if it is equal the target process PID, coreclr is ready and we don't need to wait on the "startup" semaphore.

### CODE REVIEW

**Comment by mikem8361** [comment]
Issue #5152

**Comment by mikem8361** [comment]
@jkotas, @janvorli and @noahfalk could you review? Thanks.

File: src/pal/src/thread/process.cpp [hunk 1]
```diff
@@ -1468,6 +1471,9 @@ class PAL_RuntimeStartupHelper
     // Debugger waits on this semaphore and the runtime signals it on startup.
     sem_t *m_startupSem;
 
+    // Debuggee waits on this semaphore and the debugger signals it after the callback returns.
```
Comment (noahfalk) [inline src/pal/src/thread/process.cpp]: Its probably not apparent in context here what 'the callback' refers to. Perhaps just say debugger signals this when it  is ready for the debuggee to resume executing?

File: src/pal/src/thread/process.cpp [hunk 2]
```diff
@@ -1624,10 +1667,95 @@ class PAL_RuntimeStartupHelper
...
+            pe = ERROR_MOD_NOT_FOUND;
+            goto exit;
+        }
+
+        // Load dac and call entry point to get g_coreclrProcessIsReady RVA
+        getrva = (GetRVA_PFN)dlsym(dl_handle, "GetCoreClrProcessIsReadyRVA");
+        if (getrva == NULL)
+        {
+            WARN("dlsym(GetCoreClrProcessIsReadyRVA) failed; dlerror says '%s'\n", dlerror());
+            pe = ERROR_MOD_NOT_FOUND;
```
Comment (noahfalk) [inline src/pal/src/thread/process.cpp]: Have you tested the case with new dbgshim and old coreclr/dac? If I follow the code correctly, this is the point where the back-compat handling would need to return *pCodeClrIsReady = true

File: src/pal/src/thread/process.cpp [hunk 3]
```diff
@@ -1624,10 +1667,95 @@ class PAL_RuntimeStartupHelper
...
+            dlclose(dl_handle);
+        }
+        if (hProcess != NULL)
+        {
+            CloseHandle(hProcess);
+        }
+        return pe;
+    }
+
+    PAL_ERROR InvokeStartupCallback(bool *pCoreClrIsReady)
```
Comment (noahfalk) [inline src/pal/src/thread/process.cpp]: Maybe name this InvokeStartupCallbackIfReady? Its a little confusing that it may not do what the title says it will do.

File: src/pal/src/thread/process.cpp [hunk 4]
```diff
@@ -1643,24 +1771,38 @@ class PAL_RuntimeStartupHelper
...
 
-                        // Currently only the first coreclr module in a process is supported
-                        break;
+                    // Currently only the first coreclr module in a process is supported
+                    break;
                 }
             }
 
         exit:
+            if (*pCoreClrIsReady)
```
Comment (noahfalk) [inline src/pal/src/thread/process.cpp]: Previously setting the continue semaphore was done unconditionally in StartupHelperThread and I wasn't sure what the rational was to move it here instead. I don't think we want to make setting this semaphore conditional on m_canceled right?

File: src/pal/src/thread/process.cpp [hunk 5]
```diff
@@ -1624,10 +1667,95 @@ class PAL_RuntimeStartupHelper
...
+        address = ((BYTE *)module->BaseAddress) + (*getrva)();
+
+        hProcess = OpenProcess(READ_CONTROL, FALSE, m_processId);
+        if (hProcess == NULL)
+        {
+            pe = GetLastError();
+            goto exit;
+        }
+
+        // Get the contents of g_coreclrProcessIsReady in the coreclr process. The errors from read memory
```
Comment (noahfalk) [inline src/pal/src/thread/process.cpp]: Somewhere in the comments it would be good to add a general description enumerating all the different interesting windows of time and how this algorithm handles attaching during each of them. This comment here is handy for one aspect of it, but I didn't catch anything that gave a full description.

File: src/dlls/mscoree/unixinterface.cpp [hunk 6]
```diff
@@ -353,88 +353,3 @@ int coreclr_execute_assembly(
     return hr;
 }
 
-#ifdef PLATFORM_UNIX
```
Comment (noahfalk) [inline src/dlls/mscoree/unixinterface.cpp]: Does this edit relate to the bug fix? I'm not opposed to cleanup, but I don't know where the bar is now. If this is an unrelated code change you might be asked to split it out into a seperate commit that won't get integrated to RTM.

File: src/debug/ee/debugger.cpp [hunk 7]
```diff
@@ -2113,18 +2113,9 @@ HRESULT Debugger::Startup(void)
             ShutdownTransport();
             ThrowHR(hr);
         }
-
     #ifdef FEATURE_PAL
         PAL_SetShutdownCallback(AbortTransport);
     #endif // FEATURE_PAL
-
-         bool waitForAttach = CLRConfig::GetConfigValue(CLRConfig::UNSUPPORTED_DbgWaitForDebuggerAttach) != 0;
```
Comment (noahfalk) [inline src/debug/ee/debugger.cpp]: Does this edit relate to the bug fix? I'm not opposed to cleanup, but I don't know where the bar is now. If this is an unrelated code change you might be asked to split it out into a seperate commit that won't get integrated to RTM.

File: src/pal/src/config.h.in [hunk 8]
```diff
@@ -101,6 +101,10 @@
 #cmakedefine01 PTHREAD_CREATE_MODIFIES_ERRNO
 #cmakedefine01 SEM_INIT_MODIFIES_ERRNO
 #cmakedefine01 HAVE_PROCFS_CTL
+#cmakedefine01 HAVE_PROCFS_MEM
+#cmakedefine01 HAVE_PROCFS_MAPS
+#cmakedefine01 HAVE_PROCFS_STAT
+#cmakedefine01 HAVE_PROCFS_STATUS
```
Comment (noahfalk) [inline src/pal/src/config.h.in]: Did these changes relate to the bug fix?

File: src/debug/ee/debugger.cpp (hunk 7 above)
Comment (mikem8361) [inline src/debug/ee/debugger.cpp]: This change is a lot more related than the unixinterface.cpp/ExecuteAssembly change.  It needs to be removed and related to the attach scenario.

File: src/pal/src/config.h.in (hunk 8 above)
Comment (mikem8361) [inline src/pal/src/config.h.in]: These are necessary to get ReadProcessMemory to work correctly at least HAVE_PROCFS_MEM, but the rest are all about correctness and I think related enough.

**Comment by mikem8361** [comment]
@noahfalk could you take a quick look at the code review feedback changes?

**Comment by mikem8361** [comment]
Create new PR after reworking for OSX problem in ReadProcessMemory.

**Comment by noahfalk** [comment]
lgtm
